### PR TITLE
Fix Part of #4938: Introduce Create profile screen

### PR DESCRIPTION
**Pull request by adhiamboperes**
## Explanation
Fix Part of #4938: Add a new Activity and associated Fragments and Presenters to allow a new learner to create a profile.
This does not include domain changes. 

- Learner should be able to click “Continue” if they have entered their nickname, even without adding a profile picture.
- Learner should not be able to click “Continue” if they have not entered their nickname and an error message should be displayed.
- The learner can select a profile picture.

Placeholder tests have been added to ensure navigation tests are not forgotten. These will fail once navigation has been implemented.

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
||||
| --- | --- | --- |
|| Portrait | Landscape |
|Mobile Light Mode|Screenshot_1719774012|Screenshot_1719774028|
|Tablet Dark Mode|Screenshot_1719774236|Screenshot_1719774244|

## All Tests Passing on Espresso
Screenshot 2024-05-24 at 01 42 41

### CODE REVIEW

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivity.kt
Comment (theMr17) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivity.kt]: Please import the screen name directly.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivityPresenter.kt
Comment (theMr17) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivityPresenter.kt]:
```suggestion
      ).commitNow()
```
Nit.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt
Comment (theMr17) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt]: We can rename this to better indicate the error it tracks (for eg, `hasNicknameError`).

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: Do we need to do this? It'd be much preferred to use our existing utilities to load images via Glide as, otherwise, this can complicate our Glide management.

Ditto elsewhere.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]:
```suggestion
      createProfileViewModel.hasError.set(nickname.isBlank())
```

Can be simplified a bit.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt]: Could we perhaps name it something more specific to this purpose than 'hasError'?

File: app/src/main/res/values/styles.xml
Comment (BenHenning) [inline app/src/main/res/values/styles.xml]: Is the double newline intentional?

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt:233]: Probably would be nice also having a test to verify that the error text does NOT show if the nickname isn't empty.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: Is there a way to test that an image was successfully picked & processed by the fragment presenter, as well?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @adhiamboperes! Just had a few thoughts--PTAL. I also liked to see the placeholder tests in these PRs for later functionality (to help ensure that the tests are updated for each subsequent PR once the expected functionality is added).

Could you please also make sure that this PR's description has finished being filled in?

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (theMr17) [inline app/src/main/res/layout/create_profile_fragment.xml]: Maybe we should introduce a custom style for the error text views. This can be reused in upcoming onboarding related screens as well.

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (theMr17) [inline app/src/main/res/layout/create_profile_fragment.xml]: Same as the error text view, we can introduce a custom style for this as well. Ditto for the label of this edit text. These styles can be reused in the upcoming PRs.

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (theMr17) [inline app/src/main/res/layout/create_profile_fragment.xml]: Also noticed that this title has same styling in all the other screens in the mocks. So, this can also be updated with a custom style.

File: app/src/main/res/values-night/color_palette.xml
Comment (theMr17) [inline app/src/main/res/values-night/color_palette.xml:238]: There is already a  `color_palette_error_text_color` defined in the file ([Line 58](https://github.com/oppia/oppia-android/blob/3c7713d832badacf842a5b19ec0f8b6a4ce28fba/app/src/main/res/values-night/color_palette.xml#L58)), which is similar to this one. Since, this is a different color (which is more specific  to the onboarding screens), we should name it accordingly.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (theMr17) [inline app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: It would be nice to hide the error message when there is change in text in the edit text, instead of hiding it when the user clicks on the `Continue` button again.

Since the error message says,
> Click in the box above to type your nickname.

Alternatively we could just hide the error message when the user clicks on the edit text to type the nickname (i.e., the edit text regains focus).

**Review by theMr17 — CHANGES_REQUESTED** [review]
Thanks @adhiamboperes! Left a few comments - PTAL.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: Done

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt]: I just renamed this to `hasErrorMessage` and updated the KDoc. There is more than one type of error when creating a profile e.g `ProfileNameOnlyLettersException`, And we are using the same eeror view for all.

File: app/src/main/res/values/styles.xml
Comment (adhiamboperes) [reply app/src/main/res/values/styles.xml]: No, this probably came about from resolving conflicts.

Comment (adhiamboperes) [reply app/src/main/res/values/styles.xml]: Fixed

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivity.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivity.kt]: Done

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivityPresenter.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileActivityPresenter.kt]: Fixed

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileViewModel.kt]: I just renamed this to hasErrorMessage and updated the KDoc. There is more than one type of error when creating a profile e.g ProfileNameOnlyLettersException, And we are using the same error view for all.

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (adhiamboperes) [reply app/src/main/res/layout/create_profile_fragment.xml]: Created two new styles

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (adhiamboperes) [reply app/src/main/res/layout/create_profile_fragment.xml]: I didn't observe enough identical properties between the alternate layout files to warrant a shared style, aside from font size and color.

File: app/src/main/res/values-night/color_palette.xml
Comment (adhiamboperes) [reply app/src/main/res/values-night/color_palette.xml:238]: I created anew one because the actual color values are different. The existing one is pink which is pale on a white background and may not pass accessibility scanner checks, so I created a darker version.

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (adhiamboperes) [reply app/src/main/res/layout/create_profile_fragment.xml]: Done, used the OnboardingHeaderStyle

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: Added a text watcher to listen for text change.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (adhiamboperes) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt:233]: I added these assertions.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: Switched to our utilities.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (adhiamboperes) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: I have added `testFragment_tapToAddPictureClicked_loadsTheImageFromGallery` to verify that the image was picked, but this does not cover the presenter processing it because The presenter  functionality is not public in the test.

**Comment by adhiamboperes** [comment]
@theMr17, @BenHenning, I have responded/Addressed your comments. PTAL.

File: app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt
Comment (theMr17) [inline app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt]: I have two suggestions here:
1. Rename `data: Intent?` parameter to `intent: Intent?` to enhance code readability.
2. The code can be simplified a bit as shown below. We don't need to check for `data?.let {}`, since it will be checked automatically by the `checkNotNull` block.

```suggestion
  fun handleOnActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
    if (requestCode == GALLERY_INTENT_RESULT_CODE && resultCode == Activity.RESULT_OK) {
      binding.createProfilePicturePrompt.visibility = View.GONE
      selectedImage = checkNotNull(intent?.data.toString()) { "Could not find the selected image." }
      imageLoader.loadBitmap(
        selectedImage,
        ImageViewTarget(uploadImageView)
      )
    }
  }
```

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (theMr17) [inline app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]:
```suggestion
  fun testFragment_continueButtonClicked_filledNickname_doesNotShowErrorText() {
```
Nit.

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (theMr17) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: Please add tests to verify this behavior.

File: app/src/main/res/values/styles.xml
Comment (theMr17) [inline app/src/main/res/values/styles.xml:757]:
```suggestion
  <style name="OnboardingEditTextStyle" parent="TextViewStart">
    <item name="android:paddingStart">@dimen/onboarding_shared_padding_medium</item>
    <item name="android:paddingTop">@dimen/onboarding_shared_padding_small</item>
```
Please rename the style names to something more generic. So that, these can be used in the upcoming screens of the onboarding flow. For eg, in the EditText of the Create Pin screen [here](https://www.figma.com/design/PnCtQKDR5VWbIlA3Ui7rTa/Oppia---Onboarding-Android-App?node-id=1446-17548&t=d7O8z86myfxHXMZQ-4).

Ditto for the label style as well.

File: app/src/main/res/layout-sw600dp-port/create_profile_fragment.xml
Comment (theMr17) [inline app/src/main/res/layout-sw600dp-port/create_profile_fragment.xml]: `android:autofillHints="false"` can be removed, since it will be set by the custom style.

Maybe this got skipped while introducing the custom style. Other layout files does not contain this attribute.

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (theMr17) [reply app/src/main/res/layout/create_profile_fragment.xml]: Considering that the Create Pin screen will also have same error message style [here](https://www.figma.com/design/PnCtQKDR5VWbIlA3Ui7rTa/Oppia---Onboarding-Android-App?node-id=1947-38772&t=uOusGIWdsjFK3X67-4), maybe we should create a shared style. What do you think about it?

Although, it will only consist of `fontFamily`, `textColor` & `textSize` but, in case of any modifications to the style later, this will serve as a single source for all the error messages.

**Review by theMr17 — CHANGES_REQUESTED** [review]
Thanks @adhiamboperes! I have left a few more comments - PTAL.

File: app/src/main/res/layout/create_profile_fragment.xml
Comment (adhiamboperes) [reply app/src/main/res/layout/create_profile_fragment.xml]: Done

File: app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboardingv2/CreateProfileFragmentPresenter.kt]: Added `testFragment_onTextChanged_afterError_hidesErrorMessage`

File: app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt
Comment (adhiamboperes) [reply app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt]: Done, thanks.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (adhiamboperes) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: Fixed

File: app/src/main/res/values/styles.xml
Comment (adhiamboperes) [reply app/src/main/res/values/styles.xml:757]: I'm fine with keeping it as is, since those screens are part of the onboarding flow as well.

File: app/src/main/res/layout-sw600dp-port/create_profile_fragment.xml
Comment (adhiamboperes) [reply app/src/main/res/layout-sw600dp-port/create_profile_fragment.xml]: Done

**Comment by adhiamboperes** [comment]
@theMr17, PTAL.

**Review by theMr17 — APPROVED** [review]
Thanks, @adhiamboperes, nice work here! LGTM.

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: Ditto for these tests (per my comment in your earlier PR--these can probably be tested on Robolectric by verifying Activity.isFinishing, instead).

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (BenHenning) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: Could we verify that an image load request was sent to the image loader? I think there are other tests in the codebase that verify this sort of flow using [TestGlideImageLoader](https://github.com/oppia/oppia-android/blob/develop/utility/src/main/java/org/oppia/android/util/parser/image/TestGlideImageLoader.kt).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @adhiamboperes! Had a few follow-up comments--PTAL.

Separately, are these the final mocks? The "Tap here to add picture" seems rather rough UI-wise. Is there any way we can clean this up a bit to not have hard borders on top of the profile icon?

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (adhiamboperes) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: Done

File: app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt
Comment (adhiamboperes) [reply app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt]: Thanks, I found a way to verify this.

**Comment by adhiamboperes** [comment]
Thanks for reviewing @BenHenning. 
Re - mocks: PTAL at https://www.figma.com/design/PnCtQKDR5VWbIlA3Ui7rTa/Oppia---Onboarding-Android-App?node-id=1446-13996&t=vCOOSRs3SvZiKRqi-4 

I have made code changes so that the text is not fully covering the icon, which is closer to the mocks, but I'm not sure if this satisfies the "hard borders" comment.

**Review by BenHenning — APPROVED** [review]
Thanks @adhiamboperes. Approving to unblock the PR, but re: the mocks it does look a bit rough I think. Also, reminder to remove the 'blocked' part of your title before merging this into develop. :)

@seanlip do you have any thoughts on the "tap here to add a picture" portion of the screenshots in the PR description (or the mocks linked above)? I'm wondering if we can maybe soften the borders a bit so that there isn't such an abrupt edge between the text background and the profile image layered underneath it.

Another option might be to drop the help text and use a spotlight to teach the user that it's there.

**Comment by seanlip** [comment]
@BenHenning Thanks for flagging. I took a look at the mocks and it seems that the UI here matches what's provided in the mocks. I'll bring this up with the design team and ask if any changes should be made, but I think it's still worth merging this PR now and doing that as a small fix-up later if we decide to change it.

**Comment by seanlip** [comment]
A quick update -- I did chat with the design team about this. They have a short-term suggestion and a long-term suggestion.

- Short-term suggestion (can be applied now): Drop the rectangle behind the text, make the background a bit more transparent, and make the text contrast with the background (e.g. have it be black). The rectangle on top of the face does look a bit awkward.
- Long-term suggestion: provide pre-selected avatars that the user can pick; these will be more user-friendly than a photo, especially given the target audience.

One other concern was i18n -- in German, for example, the text would be "Tippen Sie hier, um ein Bild hinzuzufügen" which is somewhat longer. Dropping the rectangle would probably help with that, too.

Please let me know what you think @adhiamboperes @BenHenning. Thanks!

(Also /cc designers @ashley-ya and @malee04 who were part of this discussion.)

**Comment by BenHenning** [comment]
Thanks much @seanlip, @ashley-ya, and @malee04! I really appreciate the quick turnaround. Just to check, when you say "make the background a bit more transparent" which background is that referring to?

@adhiamboperes do you have any thoughts on the short-term suggestion above? Do you think it's feasible & worth trying to incorporate here, or in a follow-up? FWIW I lean toward trying to do it here if the lift is small.

**Comment by seanlip** [comment]
By "background" I mean the part behind the text, i.e. the white/yellow picture (also, by "transparent" I just mean lighter). This is mostly a thought from me and is optional if the black text on top of the picture has sufficient contrast to be read clearly.

**Comment by adhiamboperes** [comment]
@seanlip, @ashley-ya, and @malee04!, @BenHenning, Here are some attempts to incorporate the feedback. Could you please let me know which direction to lean towards?

||Short Text|Long Text|  
|---|---|---|
|Light Background, regular font weight| Screenshot_1719243809| Screenshot_1719243779|
|Light Background, medium font weight| Screenshot_1719243744| Screenshot_1719243698|
|Yellow Background, regular font|Screenshot_1719243840|Screenshot_1719243863|

**Comment by seanlip** [comment]
Thanks @adhiamboperes! Personally, I like the second one best (highest contrast), followed by the third one, followed by the first.

@ashley-ya @malee04 @BenHenning WDYT?

**Comment by ashley-ya** [comment]
> @seanlip, @ashley-ya, and @malee04!, @BenHenning, Here are some attempts to incorporate the feedback. Could you please let me know which direction to lean towards?
> 	Short Text 	Long Text
> Light Background, regular font weight 	Screenshot_1719243809 	Screenshot_1719243779
> Light Background, medium font weight 	Screenshot_1719243744 	Screenshot_1719243698
> Yellow Background, regular font 	Screenshot_1719243840 	Screenshot_1719243863

I like the second option as well! the medium font with light background.

**Comment by adhiamboperes** [comment]
Fixed the layouts with "option 2", and updated PR screenshots. There are no other pending comments, so enabling auto-merge.

**Pull request merged**
